File: src/engine/value.ts
// MatFree Engine - Runtime Value System

export class RuntimeError extends Error {
  constructor(msg: string) { super(msg); this.name = 'RuntimeError' }
}

export class Matrix {
  data: number[]
  rows: number
  cols: number

  constructor(rows: number, cols: number, data?: number[]) {
    this.rows = rows; this.cols = cols
    this.data = data ?? new Array(rows * cols).fill(0)
  }

  static scalar(v: number): Matrix { return new Matrix(1, 1, [v]) }
  static zeros(r: number, c: number): Matrix { return new Matrix(r, c) }
  static ones(r: number, c: number): Matrix { return new Matrix(r, c, new Array(r * c).fill(1)) }
  static eye(n: number): Matrix { const m = new Matrix(n, n); for (let i = 0; i < n; i++) m.set(i, i, 1); return m }

  static rand(r: number, c: number): Matrix {
    const m = new Matrix(r, c)
    for (let i = 0; i < r * c; i++) m.data[i] = Math.random()
    return m
  }

  static randn(r: number, c: number): Matrix {
    const m = new Matrix(r, c)
    for (let i = 0; i < r * c; i++) {
      const u1 = Math.random(), u2 = Math.random()
      m.data[i] = Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2)
    }
    return m
  }

  static linspace(a: number, b: number, n: number): Matrix {
    const m = new Matrix(1, n)
    for (let i = 0; i < n; i++) m.data[i] = n > 1 ? a + i * (b - a) / (n - 1) : a
    return m
  }

  get(r: number, c: number): number { return this.data[r * this.cols + c] }
  set(r: number, c: number, v: number): void { this.data[r * this.cols + c] = v }
  numel(): number { return this.rows * this.cols }
  isScalar(): boolean { return this.rows === 1 && this.cols === 1 }
  scalarValue(): number { return this.data[0] }

  getWithBroadcast(r: number, c: number): number {
    const rr = this.rows === 1 ? 0 : r
    const cc = this.cols === 1 ? 0 : c
    return this.data[rr * this.cols + cc]
  }

  clone(): Matrix { return new Matrix(this.rows, this.cols, [...this.data]) }

  transpose(): Matrix {
    const m = new Matrix(this.cols, this.rows)
    for (let r = 0; r < this.rows; r++)
      for (let c = 0; c < this.cols; c++) m.set(c, r, this.get(r, c))
    return m
  }

  add(o: Matrix): Matrix {
    const [rr, cc] = this.broadcastSize(o)
    const m = new Matrix(rr, cc)
    for (let r = 0; r < rr; r++) for (let c = 0; c < cc; c++) m.set(r, c, this.getWithBroadcast(r, c) + o.getWithBroadcast(r, c))
    return m
  }

  sub(o: Matrix): Matrix {
    const [rr, cc] = this.broadcastSize(o)
    const m = new Matrix(rr, cc)
    for (let r = 0; r < rr; r++) for (let c = 0; c < cc; c++) m.set(r, c, this.getWithBroadcast(r, c) - o.getWithBroadcast(r, c))
    return m
  }

  mul(o: Matrix): Matrix {
    if (this.cols !== o.rows) throw new RuntimeError(`Matrix dimensions do not agree for multiplication: ${this.rows}x${this.cols} * ${o.rows}x${o.cols}`)
    const m = new Matrix(this.rows, o.cols)
    for (let r = 0; r < this.rows; r++)
      for (let c = 0; c < o.cols; c++) { let s = 0; for (let k = 0; k < this.cols; k++) s += this.get(r, k) * o.get(k, c); m.set(r, c, s) }
    return m
  }

  elementMul(o: Matrix): Matrix {
    const [rr, cc] = this.broadcastSize(o)
    const m = new Matrix(rr, cc)
    for (let r = 0; r < rr; r++) for (let c = 0; c < cc; c++) m.set(r, c, this.getWithBroadcast(r, c) * o.getWithBroadcast(r, c))
    return m
  }

  elementDiv(o: Matrix): Matrix {
    const [rr, cc] = this.broadcastSize(o)
    const m = new Matrix(rr, cc)
    for (let r = 0; r < rr; r++) for (let c = 0; c < cc; c++) m.set(r, c, this.getWithBroadcast(r, c) / o.getWithBroadcast(r, c))
    return m
  }

  elementPow(o: Matrix): Matrix {
    const [rr, cc] = this.broadcastSize(o)
    const m = new Matrix(rr, cc)
    for (let r = 0; r < rr; r++) for (let c = 0; c < cc; c++) m.set(r, c, Math.pow(this.getWithBroadcast(r, c), o.getWithBroadcast(r, c)))
    return m
  }

  neg(): Matrix { return new Matrix(this.rows, this.cols, this.data.map(v => -v)) }

  scalarOp(s: number, op: string): Matrix {
    return new Matrix(this.rows, this.cols, this.data.map(v => {
      switch (op) { case '+': return v + s; case '-': return v - s; case '*': return v * s; case '/': return v / s; case '^': return Math.pow(v, s); default: return v }
    }))
  }

  sum(): number { return this.data.reduce((a, b) => a + b, 0) }
  prod(): number { return this.data.reduce((a, b) => a * b, 1) }
  mean(): number { return this.sum() / this.numel() }
  minVal(): number { return Math.min(...this.data) }
  maxVal(): number { return Math.max(...this.data) }
  norm(): number { return Math.sqrt(this.data.reduce((s, v) => s + v * v, 0)) }

  det(): number {
    if (this.rows !== this.cols) throw new RuntimeError('det requires square matrix')
    const n = this.rows
    if (n === 1) return this.data[0]
    if (n === 2) return this.data[0] * this.data[3] - this.data[1] * this.data[2]
    const a = this.clone().data
    let d = 1
    for (let i = 0; i < n; i++) {
      let max = i
      for (let j = i + 1; j < n; j++) if (Math.abs(a[j * n + i]) > Math.abs(a[max * n + i])) max = j
      if (max !== i) { for (let k = 0; k < n; k++) { const t = a[i * n + k]; a[i * n + k] = a[max * n + k]; a[max * n + k] = t }; d = -d }
      if (Math.abs(a[i * n + i]) < 1e-14) return 0
      d *= a[i * n + i]
      for (let j = i + 1; j < n; j++) { const f = a[j * n + i] / a[i * n + i]; for (let k = i; k < n; k++) a[j * n + k] -= f * a[i * n + k] }
    }
    return d
  }

  inv(): Matrix {
    if (this.rows !== this.cols) throw new RuntimeError('inv requires square matrix')
    const n = this.rows
    const aug = Matrix.zeros(n, 2 * n)
    for (let r = 0; r < n; r++) { for (let c = 0; c < n; c++) aug.set(r, c, this.get(r, c)); aug.set(r, r + n, 1) }
    for (let i = 0; i < n; i++) {
      let max = i
      for (let j = i + 1; j < n; j++) if (Math.abs(aug.get(j, i)) > Math.abs(aug.get(max, i))) max = j
      if (max !== i) for (let k = 0; k < 2 * n; k++) { const t = aug.get(i, k); aug.set(i, k, aug.get(max, k)); aug.set(max, k, t) }
      const pivot = aug.get(i, i); if (Math.abs(pivot) < 1e-14) throw new RuntimeError('Matrix is singular')
      for (let k = 0; k < 2 * n; k++) aug.set(i, k, aug.get(i, k) / pivot)
      for (let j = 0; j < n; j++) if (j !== i) { const f = aug.get(j, i); for (let k = 0; k < 2 * n; k++) aug.set(j, k, aug.get(j, k) - f * aug.get(i, k)) }
    }
    const result = Matrix.zeros(n, n)
    for (let r = 0; r < n; r++) for (let c = 0; c < n; c++) result.set(r, c, aug.get(r, c + n))
    return result
  }

  trace(): number {
    if (this.rows !== this.cols) throw new RuntimeError('trace requires square matrix')
    let s = 0; for (let i = 0; i < this.rows; i++) s += this.get(i, i); return s
  }

  horzcat(o: Matrix): Matrix {
    if (this.rows !== o.rows) throw new RuntimeError('Horizontal cat: row mismatch')
    const m = new Matrix(this.rows, this.cols + o.cols)
    for (let r = 0; r < this.rows; r++) { for (let c = 0; c < this.cols; c++) m.set(r, c, this.get(r, c)); for (let c = 0; c < o.cols; c++) m.set(r, this.cols + c, o.get(r, c)) }
    return m
  }

  vertcat(o: Matrix): Matrix {
    if (this.cols !== o.cols) throw new RuntimeError('Vertical cat: col mismatch')
    const m = new Matrix(this.rows + o.rows, this.cols)
    for (let r = 0; r < this.rows; r++) for (let c = 0; c < this.cols; c++) m.set(r, c, this.get(r, c))
    for (let r = 0; r < o.rows; r++) for (let c = 0; c < this.cols; c++) m.set(this.rows + r, c, o.get(r, c))
    return m
  }

  reshape(r: number, c: number): Matrix {
    if (r * c !== this.numel()) throw new RuntimeError('reshape: numel mismatch')
    return new Matrix(r, c, [...this.data])
  }

  toString(): string {
    if (this.numel() === 0) return '[]'
    if (this.isScalar()) return formatNum(this.data[0])
    const lines: string[] = []
    for (let r = 0; r < this.rows; r++) {
      const vals: string[] = []
      for (let c = 0; c < this.cols; c++) vals.push(formatNum(this.get(r, c)).padStart(10))
      lines.push('  ' + vals.join('  '))
    }
    return lines.join('\n')
  }

  private broadcastSize(o: Matrix): [number, number] {
    const rr = Math.max(this.rows, o.rows), cc = Math.max(this.cols, o.cols)
    if (this.rows !== o.rows && this.rows !== 1 && o.rows !== 1) throw new RuntimeError('Matrix dimensions not compatible')
    if (this.cols !== o.cols && this.cols !== 1 && o.cols !== 1) throw new RuntimeError('Matrix dimensions not compatible')
    return [rr, cc]
  }
}

function formatNum(v: number): string {
  if (Number.isNaN(v)) return 'NaN'
  if (!Number.isFinite(v)) return v > 0 ? 'Inf' : '-Inf'
  if (Number.isInteger(v) && Math.abs(v) < 1e15) return v.toString()
  const s = v.toPrecision(4)
  return s.includes('.') ? s.replace(/0+$/, '').replace(/\.$/, '') : s
}

// Value types
export type FuncHandle = { type: 'builtin'; name: string } | { type: 'anonymous'; params: string[]; body: any; closure: any }
export type CellArray = { rows: number; cols: number; data: Value[] }
export type MFStruct = { [key: string]: Value }

export type ValueType = 'matrix' | 'string' | 'cell' | 'struct' | 'funcHandle' | 'logical' | 'empty'

export class Value {
  private constructor(
    public readonly type: ValueType,
    private _matrix?: Matrix,
    private _string?: string,
    private _cell?: CellArray,
    private _struct?: MFStruct,
    private _funcHandle?: FuncHandle,
  ) {}

  static fromMatrix(m: Matrix): Value { return new Value('matrix', m) }
  static fromScalar(v: number): Value { return new Value('matrix', Matrix.scalar(v)) }
  static fromString(s: string): Value { return new Value('string', undefined, s) }
  static fromLogical(b: boolean): Value { return new Value('matrix', Matrix.scalar(b ? 1 : 0)) }
  static fromCell(c: CellArray): Value { return new Value('cell', undefined, undefined, c) }
  static fromStruct(s: MFStruct): Value { return new Value('struct', undefined, undefined, undefined, s) }
  static fromFuncHandle(f: FuncHandle): Value { return new Value('funcHandle', undefined, undefined, undefined, undefined, f) }
  static empty(): Value { return new Value('empty') }

  isMatrix(): boolean { return this.type === 'matrix' }
  isString(): boolean { return this.type === 'string' }
  isCell(): boolean { return this.type === 'cell' }
  isStruct(): boolean { return this.type === 'struct' }
  isFuncHandle(): boolean { return this.type === 'funcHandle' }
  isEmpty(): boolean { return this.type === 'empty' }

  matrix(): Matrix {
    if (this._matrix) return this._matrix
    throw new RuntimeError('Value is not a matrix')
  }

  string(): string {
    if (this._string !== undefined) return this._string
    throw new RuntimeError('Value is not a string')
  }

  cell(): CellArray {
    if (this._cell) return this._cell
    throw new RuntimeError('Value is not a cell array')
  }

  struct(): MFStruct {
    if (this._struct) return this._struct
    throw new RuntimeError('Value is not a struct')
  }

  funcHandle(): FuncHandle {
    if (this._funcHandle) return this._funcHandle
    throw new RuntimeError('Value is not a function handle')
  }

  toScalar(): number {
    if (this.isMatrix()) return this.matrix().scalarValue()
    if (this.isString()) return this._string!.length === 1 ? this._string!.charCodeAt(0) : NaN
    throw new RuntimeError('Cannot convert to scalar')
  }

  toBool(): boolean {
    if (this.isMatrix()) return this.matrix().data.every(v => v !== 0)
    if (this.isString()) return this._string!.length > 0
    return false
  }

  toMatrix(): Matrix {
    if (this.isMatrix()) return this.matrix()
    if (this.isString()) {
      const d = [...this._string!].map(c => c.charCodeAt(0))
      return new Matrix(1, d.length, d)
    }
    throw new RuntimeError('Cannot convert to matrix')
  }

  display(name?: string): string {
    const label = name ?? 'ans'
    if (this.isMatrix()) {
      const m = this.matrix()
      if (m.isScalar()) return `${label} = ${formatNum(m.scalarValue())}\n`
      if (m.numel() === 0) return `${label} = []\n`
      return `${label} =\n${m.toString()}\n`
    }
    if (this.isString()) return `${label} = '${this._string}'\n`
    if (this.isStruct()) {
      const s = this._struct!
      const lines = Object.keys(s).map(k => `    ${k}: ${s[k].display().trim()}`)
      return `${label} =\n  struct with fields:\n${lines.join('\n')}\n`
    }
    if (this.isCell()) return `${label} = {${this._cell!.rows}x${this._cell!.cols} cell}\n`
    if (this.isFuncHandle()) return `${label} = @function_handle\n`
    return `${label} = [empty]\n`
  }
}

File: src/core/value.cpp
// MatFree - Runtime value and matrix implementation
// Copyright (c) 2026 MatFree Contributors - MIT License

#include "value.h"
#include <random>
#include <cassert>

namespace matfree {

// ============================================================================
// Matrix implementation
// ============================================================================

static std::mt19937& rng() {
    static std::mt19937 gen(std::random_device{}());
    return gen;
}

Matrix Matrix::rand(size_t rows, size_t cols) {
    Matrix m(rows, cols);
    std::uniform_real_distribution<double> dist(0.0, 1.0);
    for (auto& v : m.data_) v = dist(rng());
    return m;
}

Matrix Matrix::randn(size_t rows, size_t cols) {
    Matrix m(rows, cols);
    std::normal_distribution<double> dist(0.0, 1.0);
    for (auto& v : m.data_) v = dist(rng());
    return m;
}

Matrix Matrix::transpose() const {
    Matrix result(cols_, rows_);
    for (size_t i = 0; i < rows_; i++) {
        for (size_t j = 0; j < cols_; j++) {
            result(j, i) = (*this)(i, j);
        }
    }
    return result;
}

void Matrix::broadcastCheck(const Matrix& a, const Matrix& b,
                             size_t& rows, size_t& cols) {
    if (a.rows_ == b.rows_ && a.cols_ == b.cols_) {
        rows = a.rows_;
        cols = a.cols_;
    } else if (a.isScalar()) {
        rows = b.rows_;
        cols = b.cols_;
    } else if (b.isScalar()) {
        rows = a.rows_;
        cols = a.cols_;
    } else if (a.rows_ == b.rows_ && (a.cols_ == 1 || b.cols_ == 1)) {
        rows = a.rows_;
        cols = std::max(a.cols_, b.cols_);
    } else if (a.cols_ == b.cols_ && (a.rows_ == 1 || b.rows_ == 1)) {
        rows = std::max(a.rows_, b.rows_);
        cols = a.cols_;
    } else {
        throw RuntimeError("Matrix dimensions must agree (" +
            std::to_string(a.rows_) + "x" + std::to_string(a.cols_) + " vs " +
            std::to_string(b.rows_) + "x" + std::to_string(b.cols_) + ")");
    }
}

double Matrix::getWithBroadcast(size_t r, size_t c) const {
    size_t actualR = (rows_ == 1) ? 0 : r;
    size_t actualC = (cols_ == 1) ? 0 : c;
    return (*this)(actualR, actualC);
}

Matrix Matrix::operator+(const Matrix& other) const {
    size_t r, c;
    broadcastCheck(*this, other, r, c);
    Matrix result(r, c);
    for (size_t i = 0; i < r; i++)
        for (size_t j = 0; j < c; j++)
            result(i, j) = getWithBroadcast(i, j) + other.getWithBroadcast(i, j);
    return result;
}

Matrix Matrix::operator-(const Matrix& other) const {
    size_t r, c;
    broadcastCheck(*this, other, r, c);
    Matrix result(r, c);
    for (size_t i = 0; i < r; i++)
        for (size_t j = 0; j < c; j++)
            result(i, j) = getWithBroadcast(i, j) - other.getWithBroadcast(i, j);
    return result;
}

Matrix Matrix::operator*(const Matrix& other) const {
    // Matrix multiplication
    if (isScalar()) return other * scalarValue();
    if (other.isScalar()) return *this * other.scalarValue();

    if (cols_ != other.rows_) {
        throw RuntimeError("Inner matrix dimensions must agree for multiplication (" +
            std::to_string(rows_) + "x" + std::to_string(cols_) + " * " +
            std::to_string(other.rows_) + "x" + std::to_string(other.cols_) + ")");
    }

    Matrix result(rows_, other.cols_);
    for (size_t i = 0; i < rows_; i++) {
        for (size_t j = 0; j < other.cols_; j++) {
            double sum = 0.0;
            for (size_t k = 0; k < cols_; k++) {
                sum += (*this)(i, k) * other(k, j);
            }
            result(i, j) = sum;
        }
    }
    return result;
}

Matrix Matrix::elementMul(const Matrix& other) const {
    size_t r, c;
    broadcastCheck(*this, other, r, c);
    Matrix result(r, c);
    for (size_t i = 0; i < r; i++)
        for (size_t j = 0; j < c; j++)
            result(i, j) = getWithBroadcast(i, j) * other.getWithBroadcast(i, j);
    return result;
}

Matrix Matrix::elementDiv(const Matrix& other) const {
    size_t r, c;
    broadcastCheck(*this, other, r, c);
    Matrix result(r, c);
    for (size_t i = 0; i < r; i++)
        for (size_t j = 0; j < c; j++)
            result(i, j) = getWithBroadcast(i, j) / other.getWithBroadcast(i, j);
    return result;
}

Matrix Matrix::elementPow(const Matrix& other) const {
    size_t r, c;
    broadcastCheck(*this, other, r, c);
    Matrix result(r, c);
    for (size_t i = 0; i < r; i++)
        for (size_t j = 0; j < c; j++)
            result(i, j) = std::pow(getWithBroadcast(i, j), other.getWithBroadcast(i, j));
    return result;
}

Matrix Matrix::operator-() const {
    Matrix result(rows_, cols_);
    for (size_t i = 0; i < data_.size(); i++)
        result.data_[i] = -data_[i];
    return result;
}

Matrix Matrix::operator+(double s) const {
    Matrix result(rows_, cols_);
    for (size_t i = 0; i < data_.size(); i++)
        result.data_[i] = data_[i] + s;
    return result;
}

Matrix Matrix::operator-(double s) const {
    Matrix result(rows_, cols_);
    for (size_t i = 0; i < data_.size(); i++)
        result.data_[i] = data_[i] - s;
    return result;
}

Matrix Matrix::operator*(double s) const {
    Matrix result(rows_, cols_);
    for (size_t i = 0; i < data_.size(); i++)
        result.data_[i] = data_[i] * s;
    return result;
}

Matrix Matrix::operator/(double s) const {
    Matrix result(rows_, cols_);
    for (size_t i = 0; i < data_.size(); i++)
        result.data_[i] = data_[i] / s;
    return result;
}

Matrix Matrix::power(double s) const {
    Matrix result(rows_, cols_);
    for (size_t i = 0; i < data_.size(); i++)
        result.data_[i] = std::pow(data_[i], s);
    return result;
}

// Comparison operations
Matrix Matrix::eq(const Matrix& other) const {
    size_t r, c; broadcastCheck(*this, other, r, c);
    Matrix result(r, c);
    for (size_t i = 0; i < r; i++)
        for (size_t j = 0; j < c; j++)
            result(i, j) = (getWithBroadcast(i, j) == other.getWithBroadcast(i, j)) ? 1.0 : 0.0;
    return result;
}

Matrix Matrix::ne(const Matrix& other) const {
    size_t r, c; broadcastCheck(*this, other, r, c);
    Matrix result(r, c);
    for (size_t i = 0; i < r; i++)
        for (size_t j = 0; j < c; j++)
            result(i, j) = (getWithBroadcast(i, j) != other.getWithBroadcast(i, j)) ? 1.0 : 0.0;
    return result;
}

Matrix Matrix::lt(const Matrix& other) const {
    size_t r, c; broadcastCheck(*this, other, r, c);
    Matrix result(r, c);
    for (size_t i = 0; i < r; i++)
        for (size_t j = 0; j < c; j++)
            result(i, j) = (getWithBroadcast(i, j) < other.getWithBroadcast(i, j)) ? 1.0 : 0.0;
    return result;
}

Matrix Matrix::gt(const Matrix& other) const {
    size_t r, c; broadcastCheck(*this, other, r, c);
    Matrix result(r, c);
    for (size_t i = 0; i < r; i++)
        for (size_t j = 0; j < c; j++)
            result(i, j) = (getWithBroadcast(i, j) > other.getWithBroadcast(i, j)) ? 1.0 : 0.0;
    return result;
}

Matrix Matrix::le(const Matrix& other) const {
    size_t r, c; broadcastCheck(*this, other, r, c);
    Matrix result(r, c);
    for (size_t i = 0; i < r; i++)
        for (size_t j = 0; j < c; j++)
            result(i, j) = (getWithBroadcast(i, j) <= other.getWithBroadcast(i, j)) ? 1.0 : 0.0;
    return result;
}

Matrix Matrix::ge(const Matrix& other) const {
    size_t r, c; broadcastCheck(*this, other, r, c);
    Matrix result(r, c);
    for (size_t i = 0; i < r; i++)
        for (size_t j = 0; j < c; j++)
            result(i, j) = (getWithBroadcast(i, j) >= other.getWithBroadcast(i, j)) ? 1.0 : 0.0;
    return result;
}

// Reductions
double Matrix::sum() const {
    double s = 0;
    for (auto v : data_) s += v;
    return s;
}

double Matrix::prod() const {
    double p = 1;
    for (auto v : data_) p *= v;
    return p;
}

double Matrix::mean() const {
    return sum() / static_cast<double>(numel());
}

double Matrix::minVal() const {
    if (data_.empty()) throw RuntimeError("Cannot find min of empty matrix");
    return *std::min_element(data_.begin(), data_.end());
}

double Matrix::maxVal() const {
    if (data_.empty()) throw RuntimeError("Cannot find max of empty matrix");
    return *std::max_element(data_.begin(), data_.end());
}

double Matrix::norm(double p) const {
    if (p == 2.0 && isVector()) {
        double s = 0;
        for (auto v : data_) s += v * v;
        return std::sqrt(s);
    }
    if (p == 1.0) {
        double s = 0;
        for (auto v : data_) s += std::abs(v);
        return s;
    }
    if (std::isinf(p)) {
        double mx = 0;
        for (auto v : data_) mx = std::max(mx, std::abs(v));
        return mx;
    }
    // General p-norm for vectors
    double s = 0;
    for (auto v : data_) s += std::pow(std::abs(v), p);
    return std::pow(s, 1.0 / p);
}

Matrix Matrix::sumAlongDim(int dim) const {
    if (dim == 1) {
        // Sum along rows (result is 1 x cols)
        Matrix result(1, cols_);
        for (size_t j = 0; j < cols_; j++) {
            double s = 0;
            for (size_t i = 0; i < rows_; i++) s += (*this)(i, j);
            result(0, j) = s;
        }
        return result;
    } else {
        // Sum along columns (result is rows x 1)
        Matrix result(rows_, 1);
        for (size_t i = 0; i < rows_; i++) {
            double s = 0;
            for (size_t j = 0; j < cols_; j++) s += (*this)(i, j);
            result(i, 0) = s;
        }
        return result;
    }
}

Matrix Matrix::meanAlongDim(int dim) const {
    Matrix s = sumAlongDim(dim);
    double divisor = (dim == 1) ? static_cast<double>(rows_) : static_cast<double>(cols_);
    for (auto& v : s.data()) v /= divisor;
    return s;
}

Matrix Matrix::getRow(size_t row) const {
    Matrix result(1, cols_);
    for (size_t j = 0; j < cols_; j++) result(0, j) = (*this)(row, j);
    return result;
}

Matrix Matrix::getCol(size_t col) const {
    Matrix result(rows_, 1);
    for (size_t i = 0; i < rows_; i++) result(i, 0) = (*this)(i, col);
    return result;
}

Matrix Matrix::submatrix(size_t r1, size_t c1, size_t r2, size_t c2) const {
    size_t nr = r2 - r1 + 1, nc = c2 - c1 + 1;
    Matrix result(nr, nc);
    for (size_t i = 0; i < nr; i++)
        for (size_t j = 0; j < nc; j++)
            result(i, j) = (*this)(r1 + i, c1 + j);
    return result;
}

void Matrix::setRow(size_t row, const Matrix& vals) {
    for (size_t j = 0; j < cols_ && j < vals.numel(); j++)
        (*this)(row, j) = vals(j);
}

void Matrix::setCol(size_t col, const Matrix& vals) {
    for (size_t i = 0; i < rows_ && i < vals.numel(); i++)
        (*this)(i, col) = vals(i);
}

Matrix Matrix::reshape(size_t newRows, size_t newCols) const {
    if (newRows * newCols != numel()) {
        throw RuntimeError("Cannot reshape " + std::to_string(rows_) + "x" +
            std::to_string(cols_) + " to " + std::to_string(newRows) + "x" +
            std::to_string(newCols));
    }
    return Matrix(newRows, newCols, data_);
}

Matrix Matrix::horzcat(const std::vector<Matrix>& matrices) {
    if (matrices.empty()) return Matrix();
    size_t rows = matrices[0].rows();
    size_t totalCols = 0;
    for (auto& m : matrices) {
        if (m.rows() != rows && !m.isEmpty())
            throw RuntimeError("Dimensions of arrays being concatenated are not consistent");
        totalCols += m.cols();
    }

    Matrix result(rows, totalCols);
    size_t colOffset = 0;
    for (auto& m : matrices) {
        for (size_t i = 0; i < m.rows(); i++)
            for (size_t j = 0; j < m.cols(); j++)
                result(i, colOffset + j) = m(i, j);
        colOffset += m.cols();
    }
    return result;
}

Matrix Matrix::vertcat(const std::vector<Matrix>& matrices) {
    if (matrices.empty()) return Matrix();
    size_t cols = matrices[0].cols();
    size_t totalRows = 0;
    for (auto& m : matrices) {
        if (m.cols() != cols && !m.isEmpty())
            throw RuntimeError("Dimensions of arrays being concatenated are not consistent");
        totalRows += m.rows();
    }

    Matrix result(totalRows, cols);
    size_t rowOffset = 0;
    for (auto& m : matrices) {
        for (size_t i = 0; i < m.rows(); i++)
            for (size_t j = 0; j < m.cols(); j++)
                result(rowOffset + i, j) = m(i, j);
        rowOffset += m.rows();
    }
    return result;
}

std::string Matrix::toString() const {
    std::ostringstream oss;
    display(oss);
    return oss.str();
}

void Matrix::display(std::ostream& os, const std::string& name) const {
    if (!name.empty()) {
        os << name << " =" << std::endl << std::endl;
    }

    if (isEmpty()) {
        os << "     []" << std::endl;
        return;
    }

    if (isScalar()) {
        os << "   " << std::setprecision(4) << data_[0] << std::endl;
        return;
    }

    // Determine formatting
    bool allIntegers = true;
    double maxAbs = 0;
    for (auto v : data_) {
        if (v != std::floor(v)) allIntegers = false;
        maxAbs = std::max(maxAbs, std::abs(v));
    }

    int width = 10;
    int precision = 4;
    if (allIntegers && maxAbs < 1e6) {
        width = static_cast<int>(std::to_string(static_cast<long long>(maxAbs)).size()) + 5;
        precision = 0;
    }

    for (size_t i = 0; i < rows_; i++) {
        os << "   ";
        for (size_t j = 0; j < cols_; j++) {
            if (allIntegers && maxAbs < 1e6) {
                os << std::setw(width) << static_cast<long long>((*this)(i, j));
            } else {
                os << std::setw(width) << std::setprecision(precision)
                   << std::fixed << (*this)(i, j);
            }
        }
        os << std::endl;
    }
}

// ============================================================================
// Value implementation
// ============================================================================

bool Value::toBool() const {
    switch (type_) {
        case ValueType::MATRIX:
        case ValueType::LOGICAL: {
            // MATLAB: all elements must be nonzero
            for (size_t i = 0; i < matrix_.numel(); i++) {
                if (matrix_(i) == 0.0) return false;
            }
            return matrix_.numel() > 0;
        }
        case ValueType::STRING:
            return !string_.empty();
        default:
            throw RuntimeError("Cannot convert value to logical");
    }
}

Matrix Value::toMatrix() const {
    switch (type_) {
        case ValueType::MATRIX:
        case ValueType::LOGICAL:
            return matrix_;
        case ValueType::STRING: {
            // Convert string to array of character codes
            Matrix m(1, string_.size());
            for (size_t i = 0; i < string_.size(); i++)
                m(0, i) = static_cast<double>(string_[i]);
            return m;
        }
        default:
            throw RuntimeError("Cannot convert to numeric matrix");
    }
}

std::string Value::toString() const {
    std::ostringstream oss;
    switch (type_) {
        case ValueType::MATRIX:
        case ValueType::LOGICAL:
            oss << matrix_.toString();
            break;
        case ValueType::STRING:
            oss << "'" << string_ << "'";
            break;
        case ValueType::CELL_ARRAY:
            oss << "{" << cellArray_.rows << "x" << cellArray_.cols << " cell}";
            break;
        case ValueType::STRUCT:
            oss << "struct with " << struct_.fields.size() << " fields";
            break;
        case ValueType::FUNC_HANDLE:
            oss << "@" << funcHandle_.name;
            break;
        case ValueType::EMPTY:
            oss << "[]";
            break;
        default:
            oss << "<unknown>";
            break;
    }
    return oss.str();
}

void Value::display(std::ostream& os, const std::string& name) const {
    switch (type_) {
        case ValueType::MATRIX:
        case ValueType::LOGICAL:
            if (!name.empty()) os << name << " =" << std::endl << std::endl;
            matrix_.display(os);
            os << std::endl;
            break;
        case ValueType::STRING:
            if (!name.empty()) os << name << " =" << std::endl << std::endl;
            os << "    '" << string_ << "'" << std::endl << std::endl;
            break;
        case ValueType::CELL_ARRAY:
            if (!name.empty()) os << name << " =" << std::endl << std::endl;
            os << "  {" << cellArray_.rows << "x" << cellArray_.cols << " cell}" << std::endl << std::endl;
            break;
        case ValueType::STRUCT:
            if (!name.empty()) os << name << " =" << std::endl << std::endl;
            os << "  struct with fields:" << std::endl;
            for (auto& [k, v] : struct_.fields) {
                os << "    " << k << ": ";
                if (v) os << v->toString();
                else os << "[]";
                os << std::endl;
            }
            os << std::endl;
            break;
        case ValueType::FUNC_HANDLE:
            if (!name.empty()) os << name << " =" << std::endl << std::endl;
            os << "    @" << funcHandle_.name << std::endl << std::endl;
            break;
        case ValueType::EMPTY:
            if (!name.empty()) os << name << " =" << std::endl << std::endl;
            os << "     []" << std::endl << std::endl;
            break;
        default:
            break;
    }
}

} // namespace matfree
